### ISSUE TITLE
Duplicate associations in summary statistics files

### ISSUE THREAD

**Issue by earlEBI** [EBISPOT/goci#815]
This issue was brought to our attention by Daniel Suvages via gwas-info so we should keep them in the loop. 
Copy of the e-mail:

I have been investigating a user report about duplication in our datasets. When digging into the data, I have identified a range of issues in our pipelines, however I also noticed that some of the duplications are coming from the GWAS Catalog (potentially hundreds of summary stats are affected, on our side its number is more than 500). Please let me provide some details in the example of study GCST003395 [1].
 
The raw data: contains 9_583_642 rows, but only 9_558_711 unique, meaning there are 24_931 rows with identical values in all fields [2]. When calculating the number of unique variants (CHR, POS_b37, A1, A2) I got 9_558_444, meaning there are 267 duplicated variants with different statistical parameters [3]. Based on the study page on the GWAS Catalog website, I think this level of duplication cannot be explained by potential conditional analysis (also, the p-values aren’t justify a conditional test). Based on my experience such events can happen when coordinates from an older build is mapped to a newer one. Eg. Multiple coordinates on b36 could be mapped to a single b37 coordinate (but this is just an assumption).
In the harmonized dataset these duplications propagated: 9_536_338 total, 9_523_572 unique and 9_468_100 unique variants.
 
I’m wondering if you are already aware of such duplications? As I mentioned earlier this might be quite prevalent issue. Tacking the pure duplications is trivial: just select distinct rows before saving the data, however not sure what the best way of handling the second group, when seemingly the same variant has two or more association statistics. Any insight would be highly welcomed.
 
Best,
Daniel
 
 
[1] Files sourced from:
- Raw summary stats: ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST003001-GCST004000/GCST003395/results.UKBiobank_9millionSNPs.parents_lifespan.Pilling_et_al_2016_top_1_percent.txt
- Harmonized summary stats: ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST003001-GCST004000/GCST003395/harmonised/27015805-GCST003395-EFO_0007796.h.tsv.gz
- Formatted summary stats with b37 coordinates: ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST003001-GCST004000/GCST003395/harmonised/27015805-GCST003395-EFO_0007796-build37.f.tsv.gz
 
![Screenshot 2022-08-05 at 09.58.44.png](https://images.zenhubusercontent.com/5e95cce7cc796e84ca614069/3d21e0dd-3873-4459-9d84-cb7ee242ff17)

**Comment by jdhayhurst** [comment]
I've done a bit of an investigation into the summary data from GCST003395.

In the raw data, I broke it down as follows (counting every instance of the duplicates i.e. with the pandas duplicated(keep=False) and drop_dupliactes(keep=False) methods:


9583642    total records
9534587    unique records
49055    duplicate records (all duplicates counted)
9534054    unique records based on the variant data alone
49588    duplicate records based on the variant data alone (all duplicates counted)
533    subset of above where stats differ


And for the harmonised data:

9481847    total (filtered on those with hm_variant_id) records
9456557    unique records
25290    duplicate records (all duplicates counted)
9454563    unique records based on the variant data alone
27284    duplicate records based on the variant data alone (all duplicates counted)
1994    subset of above where stats differ


Initially, this is alarming because the number of duplicates with differing association stats almost 4X for the harmonised data. Here are a four examples of these (the pairing column shows you that record 'a' from the raw file corresponds with record 'a' from the harmonised file):
```
+---------+----------------+----------+-------------------+-----+-----------+------------------+-----------------+--------------+---------------+--------------+-----------+-------------+------------+-------+
| example | raw/harmonised | pairings |    variant_id     | chr |    pos    | hm_effect_allele | hm_other_allele |   hm_beta    | effect_allele | other_allele |   freq    |    beta     |     se     |   p   |
+---------+----------------+----------+-------------------+-----+-----------+------------------+-----------------+--------------+---------------+--------------+-----------+-------------+------------+-------+
|    1    |      raw       |    a     | 10:102209775_C_CT | 10  | 102209775 |       nan        |       nan       |     nan      |       C       |      CT      | 0.979015  | 0.00117554  | 0.00303526 |  0.7  |
|    1    |      raw       |    b     |   Affx-2383868    | 10  | 102209775 |       nan        |       nan       |     nan      |      CT       |      C       | 0.0206439 | -0.0017903  | 0.00299839 | 0.55  |
|    1    |      harm      |    a     |    rs11413654     | 10  | 100450018 |        CT        |        C        | -0.00117554  |       C       |      CT      | 0.979015  | 0.00117554  | 0.00303526 |  0.7  |
|    1    |      harm      |    b     |    rs11413654     | 10  | 100450018 |        CT        |        C        | -0.00179031  |      CT       |      C       | 0.0206439 | -0.0017903  | 0.00299839 | 0.55  |
+---------+----------------+----------+-------------------+-----+-----------+------------------+-----------------+--------------+---------------+--------------+-----------+-------------+------------+-------+
|    2    |      raw       |    c     |    rs376454274    | 10  | 17787983  |       nan        |       nan       |     nan      |       A       |      G       |  0.96859  | 0.00290512  | 0.00335139 | 0.39  |
|    2    |      raw       |    d     |    rs145363125    | 10  | 18034947  |       nan        |       nan       |     nan      |       A       |      G       | 0.606068  | -0.0007883  | 0.00103352 | 0.45  |
|    2    |      harm      |    c     |    rs34863391     | 10  | 17745984  |        G         |        A        | -0.00290512  |       A       |      G       |  0.96859  | 0.00290512  | 0.00335139 | 0.39  |
|    2    |      harm      |    d     |    rs34863391     | 10  | 17745984  |        G         |        A        | 0.000788316  |       A       |      G       | 0.606068  | -0.0007883  | 0.00103352 | 0.45  |
+---------+----------------+----------+-------------------+-----+-----------+------------------+-----------------+--------------+---------------+--------------+-----------+-------------+------------+-------+
|    3    |      raw       |    e     | 1:100194200_C_CAA |  1  | 100194200 |       nan        |       nan       |     nan      |       C       |     CAA      | 0.980231  | -0.00150877 | 0.00318465 | 0.64  |
|    3    |      raw       |    f     |   Affx-52345695   |  1  | 100194200 |       nan        |       nan       |     nan      |      CAA      |      C       | 0.0197579 | 0.000864545 | 0.00306155 | 0.78  |
|    3    |      harm      |    e     |    rs149243286    |  1  | 99728644  |       CAA        |        C        |  0.00150877  |       C       |     CAA      | 0.980231  | -0.00150877 | 0.00318465 | 0.64  |
|    3    |      harm      |    f     |    rs149243286    |  1  | 99728644  |       CAA        |        C        | 0.000864545  |      CAA      |      C       | 0.0197579 | 0.000864545 | 0.00306155 | 0.78  |
+---------+----------------+----------+-------------------+-----+-----------+------------------+-----------------+--------------+---------------+--------------+-----------+-------------+------------+-------+
|    4    |      raw       |    g     |    rs374118515    |  7  | 74406888  |       nan        |       nan       |     nan      |       C       |      T       | 0.866901  | 0.000474165 | 0.00135053 | 0.73  |
|    4    |      raw       |    h     |    rs376844849    |  7  | 74839540  |       nan        |       nan       |     nan      |       G       |      A       | 0.964407  | 0.00514563  | 0.00292601 | 0.079 |
|    4    |      harm      |    g     |    rs376844849    |  7  | 74992717  |        T         |        C        | -0.000474165 |       C       |      T       | 0.866901  | 0.000474165 | 0.00135053 | 0.73  |
|    4    |      harm      |    h     |    rs376844849    |  7  | 74992717  |        T         |        C        | -0.00514563  |       G       |      A       | 0.964407  | 0.00514563  | 0.00292601 | 0.079 |
+---------+----------------+----------+-------------------+-----+-----------+------------------+-----------------+--------------+---------------+--------------+-----------+-------------+------------+-------+
```
What I'm concerned about is the introduction of duplicates from the harmonisation process. It's important to note that if the raw file is reported against a genome build that is not 38, the first step is to bring those variants over to 38 before the strand orientation. It achieves this by doing the following (pseudocode):

```
if variant_id is in ensembl:
        update position to 38
else:
        liftover from current build to 38
```

In all of the above examples we started (in the raw data) with 2 differently described variants and after harmonisation ended up with 2 identically described variants with different stats.


Example 1:

(a) and (b) have no variant IDs so both get liftover treatments. They are the same position, so receive the same 38 position. Once harmonised to the same reference the alleles because the same and beta for (a) is flipped.

So these appear to be the same variant but reported in different directions - this was a bit hard to see until harmonisation. You can see that the allele frequencies nearly sum to 1 and the betas are almost opposites (se and p are close), so I think it's possible that the author merged datasets from more than one array - I don't know how else to explain this. Do you have any ideas?


Example 2:

rs376454274 (c) is synonymous to rs34863391, which has position 17745984 in Ensembl. rs145363125 (d) is not in Ensembl 38; so liftover from 37 to 38 would have also resulted in: 17745984.

So (c) and (d) end up appearing as the same variant but I suspect that this is an error/artefact of liftover because they appear to be genuinely different variants. I don't know how to address this at the moment, happy to hear any suggestions.


Example 3:

Same as example 1.


Example 4:

Same as example 2.

rs374118515 (g) maps to 74992717 in Ensembl. rs376844849 (h) does not map to Ensembl 38, so liftover from 37 to 38 would have resulted in: 74992717 (it maps to 74992717 in dbSNP build 108).


To summarise, I think we have a combination of sources of the duplicates:

1) duplicates where present in the raw data (we can validate for this in future)

2) the raw data might be derived from more than one analysis where the same variant can be described in different ways between these analyses.

3) liftover can introduce issues

4) systematic errors or typos? (probably occur but not seen here)

5) coordinate systems?

It might make some of the issues more transparent if we annotated the records to show whether they were mapped via rsID in Ensembl or using liftover.
